docs/source/onedal/algorithms/ensembles/decision-forest.rst

```diff
@@ -66,100 +65,206 @@ size :math:`p`, their non-negative observation weights :math:`W=\{w_1,\ldots,w_n
-engine-specific and depends on the parallelization technique used
-as defined earlier:
-
-- Family: the updated state is the set of states that represent
-  individual engines in the family.
-
-- Leapfrog: the updated state is the state of the sequence with
-  the rightmost position on the sequence. The example below
-  demonstrates the idea for case of 2 subsequences (‘x’ and ‘o’)
-  of the random number sequence:
-
-- SkipAhead: the updated state is the state of the independent
-  sequence with the rightmost position on the sequence. The
-  example below demonstrates the idea for case of 2 subsequences
-  (‘x’ and ‘o’) of the random number sequence:
+If maximal number of leaf nodes equals zero, a :ref:`decision tree <dt>` is built using depth-first strategy.
+In each terminal node :math:`t`, the following recursive procedure is applied: 
+
+- Stop if the termination criteria are met.
+- Choose randomly without replacement :math:`m` feature indices :math:`J_t \in \{0, 1, \ldots, p-1\}`.
+- For each :math:`j \in J_t`, find the best split :math:`s_{j,t}` that
+  partitions subset :math:`D_t` and maximizes impurity decrease
+  :math:`\Delta i(t)`.
+- A node is a split if this split induces a decrease of the impurity greater than or equal to the predefined value.
+  Get the best split :math:`s_t` that maximizes impurity decrease
+  :math:`\Delta i` in all :math:`s_{j,t}` splits.
+- Apply this procedure recursively to :math:`t_L` and :math:`t_R`.
+
+.. _df_t_best_first_strategy:
+
+Best-first Strategy
+~~~~~~~~~~~~~~~~~~~
+
+If maximal number of leaf nodes is positive, a :ref:`decision tree <dt>` is built using best-first strategy.
+In each terminal node :math:`t`, the following steps are applied:
+
+- Stop if the termination criteria are met.
+- Choose randomly without replacement :math:`m` feature indices :math:`J_t \in \{0, 1, \ldots, p-1\}`.
+- For each :math:`j \in J_t`, find the best split :math:`s_{j,t}` that
+  partitions subset :math:`D_t` and maximizes impurity decrease
+  :math:`\Delta i(t)`.
+- A node is a split if this split induces a decrease of the impurity greater than or equal to the predefined value
+  and the number of split nodes is less or equal to :math:`\mathrm{maxLeafNodes} – 1`.
+  Get the best split :math:`s_t` that maximizes impurity decrease
+  :math:`\Delta i` in all :math:`s_{j,t}` splits.
+- Put a node into a sorted array, where sort criterion is the improvement in impurity :math:`\Delta i(t)|D_t|`.
+  The node with maximal improvement is the first in the array. For a leaf node, the improvement in impurity is zero.
+- Apply this procedure to :math:`t_L` and :math:`t_R` and grow a tree one by one getting the first element from the array
+  until the array is empty.
+
+.. _df_i_math:
+
+Inference
+---------
+
+Given decision forest classification or regression model and vectors :math:`x_1, \ldots, x_r`,
+the problem is to calculate the responses for those vectors. 
+
+.. _df_i_math_dense:
+.. _df_i_math_hist:
```

Comment:
It isn't user-friendly approach and bad user experience

Reply:
fixed